STEP2: 예외처리 및 리팩터링

eottabom opened the pull request:
1주차 STEP 2 인데,, 이제야 commit title 과 branch name 이 잘못되었다는 것을 인지했네요..ㅎㅎ

eottabom committed:
STEP1: 1단계 예약 기능 구현

- test : 메서드 이름 명확하게 수정

eottabom committed:
STEP1: 1단계 예약 기능 구현

- fix : Reservation id 에 대해서 equals, hashCode Override 추가

eottabom committed:
STEP1: 2단계 테마 관리 기능 추가

- fix: html 내용 수정 및 sql 수정

eottabom committed:
STEP1: 2단계 테마 관리 기능 추가

- feat: 테마 관리 기능에 필요한 js, html 추가

eottabom committed:
STEP1: 2단계 테마 관리 기능 추가

- feat: 테마 관리 기능 추가

eottabom committed:
STEP1: 2단계 테마 관리 기능 추가

- feat: 예외 처리 추가

Hyeon9mak commented on src/main/java/roomescape/controller/dto/ReservationRequest.java:
```diff
@@ -1,5 +1,29 @@
+				(reservationDate.isEqual(LocalDate.now()) && reservationTime.isBefore(LocalTime.now()))) {
+			throw new RoomEscapeException(ErrorCode.PAST_RESERVATION);
+		}
+	}
```
name, date 필드에 대한 null 검사 👍 요러한 검사는 presentation 계층에서 검사해주기 좋죠 😄

Hyeon9mak replied:
`reservationDate.isBefore(LocalDate.now()` 는 표현보다는 방탈출 비즈니스 로직에 가까워보이는데요!
유근님의 동료 개발자분께 "예약 가능시간을 최소 일주일 전으로 변경해주세요." 라는 요구사항이 전달 된다면, 
동료 개발자분은 controller 와 domain 패키지 중 어떤 패키지를 먼저 탐색하게 될까요? 😉

eottabom replied:
domain 패키지를 먼저 확인할 것 같아요~

https://github.com/next-step/spring-roomescape-reservation/pull/70/commits/b883977ed600b7bb2d2028023aec0cf0de23a971#diff-7237a8746b2304f7aec1072a497c9255583de52c6835de79e0ff0574aeb79697R37

controller 에서 null 체크 검사하게 수정하고,

그 외에 비즈니스 로직에 대한 예외처리를 수행하게 수정했어요~
https://github.com/next-step/spring-roomescape-reservation/pull/70/commits/b883977ed600b7bb2d2028023aec0cf0de23a971#diff-a08e170bf90a7f2fae99eafa05dfd10de6889b286503b78852bda8c78832cb3dR66

Hyeon9mak commented on src/main/java/roomescape/domain/Reservation.java:
```diff
@@ -46,6 +50,31 @@ public void setTime(ReservationTime time) {
+		this.theme = theme;
+	}
+
+	@Override
+	public boolean equals(Object o) {
+		if (this == o) {
+			return true;
+		}
+		if (o == null || getClass() != o.getClass()) {
+			return false;
+		}
+		Reservation that = (Reservation) o;
+		return Objects.equals(this.id, that.id);
+	}
+
+	@Override
```
😉 👍

Hyeon9mak commented on src/main/java/roomescape/domain/Reservation.java:
```diff
@@ -46,6 +50,31 @@ public void setTime(ReservationTime time) {
 		this.time = time;
 	}
 
+	public Theme getTheme() {
+		return this.theme;
+	}
+
+	public void setTheme(Theme theme) {
+		this.theme = theme;
+	}
```
현재 builder 를 통해 생성 단계에서 theme 를 지정해주고 있으므로,
불필요한 setter 는 제거해줄 수 있겠어요!

eottabom replied:
https://github.com/next-step/spring-roomescape-reservation/pull/70/commits/b883977ed600b7bb2d2028023aec0cf0de23a971#diff-cf4b525c481f9af0c8ecf9ad0bcc44afa76271a5e7f62987fa71434ef282e6dcL31 불필요한 setter 제거했어요~

Hyeon9mak commented on src/main/java/roomescape/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,43 @@
+import org.springframework.web.context.request.WebRequest;
+import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler;
+
+@RestControllerAdvice
+public class GlobalExceptionHandler extends ResponseEntityExceptionHandler {
+
+	private static final Logger logger = LoggerFactory.getLogger(GlobalExceptionHandler.class);
+
+	@ExceptionHandler(RoomEscapeException.class)
+	public ProblemDetail handleRoomEscapeExceptions(RoomEscapeException ex, WebRequest request) {
+		ErrorCode errorCode = ex.getErrorCode();
+		HttpStatus status = getStatusForErrorCode(errorCode);
+
+		ProblemDetail problemDetail = ProblemDetail.forStatusAndDetail(status, errorCode.getMessage());
+		problemDetail.setTitle("API Error");
+		problemDetail.setInstance(URI.create(request.getDescription(false)));
+		problemDetail.setProperty("errorCode", errorCode.getCode());
+
+		logger.error("Error occurred: {}", problemDetail, ex);
+
+		return problemDetail;
+	}
```
RoomEscapeException 외 다른 예외가 발생했을 때도 관리해주는 건 어떨까요?!

eottabom replied:
https://github.com/next-step/spring-roomescape-reservation/pull/70/commits/b883977ed600b7bb2d2028023aec0cf0de23a971#diff-ce52ea65d195a4a344d7e0955ee2036bf04e2c78e27437a42ff851d77f03bd97R35

추가했어요~

Hyeon9mak commented on src/main/java/roomescape/service/ReservationTimeService.java:
```diff
@@ -33,11 +35,23 @@ public List<ReservationTimeResponse> getReservationTimes() {
-		this.reservationTimeRepository.delete(id);
+		var isExist = this.reservationTimeRepository.isExistId(id);
+		if (isExist) {
+			this.reservationTimeRepository.delete(id);
+		}
+		else {
+			throw new RoomEscapeException(ErrorCode.NOT_FOUND_RESERVATION_TIME);
+		}
 	}
 
 	public ReservationTime getReservationTimeById(long id) {
-		return this.reservationTimeRepository.findById(id);
+//		try {
+			return this.reservationTimeRepository.findById(id);
+//		}
+//		catch (Exception ex) {
+//			throw new RoomEscapeException(ErrorCode.NOT_FOUND_RESERVATION_TIME);
+//		}
+
```
불필요해진 주석은 제거해주셔도 좋을거 같아요 😄 👍

eottabom replied:
https://github.com/next-step/spring-roomescape-reservation/pull/70/commits/b883977ed600b7bb2d2028023aec0cf0de23a971#diff-ef1662385ec72c3f06f1732502738c0bd1de54de75e17e25fe078a7552b39b91L48 제거했어요~ ㅎㅎ

Hyeon9mak commented on src/main/resources/application.properties:
```diff
@@ -2,6 +2,9 @@
 spring.datasource.driver-class-name=org.h2.Driver
 spring.datasource.url=jdbc:h2:mem:database
 spring.h2.console.enabled=true
+spring.sql.init.mode=always
+spring.sql.init.platform=h2
+spring.sql.init.schema-locations=classpath:schema.sql
```
schema-locations 를 이용한 초기화! 👍

Hyeon9mak reviewed (changes requested):
안녕하세요 유근님! 2단계도 잘 진행해주셨습니다 👍 
이전 단계에서 상세하게 설명해주신 덕분에 저도 많이 배울 수 있었습니다! 😄 👍 

이미 스프링에 익숙하신 덕택에, 잘 구현해주셔서 코멘트를 남길 부분이 많지 않은데요,
간단한 코멘트 몇 가지만 확인해주세요!

eottabom committed:
STEP2: 2단계 테마 관리 기능 추가

- fix : validateReservation 수정 및 분리
- fix : 불필요한 setter 제거
- fix : RoomEscapeException 이외의 다른 예외 처리 관리 추가
- fix : 불필요한 주석 제거

Hyeon9mak commented on src/test/java/roomescape/MissionStepTests.java:
```diff
@@ -94,4 +100,24 @@ private void createReservationTime() {
 			.body("id", is(1));
 	}
 
+	private void createTheme() {
+		Map<String, String> params = new HashMap<>();
+		params.put("name", "테마1");
+		params.put("description", "첫번째테마");
+		params.put("thumbnail", "테마이미지");
+
+		RestAssured.given()
+				.log()
+				.all()
+				.contentType(ContentType.JSON)
+				.body(params)
+				.when()
+				.post("/themes")
+				.then()
+				.log()
+				.all()
+				.statusCode(200)
+				.body("id", is(1));
+	}
```
테마 생성용 메서드를 따로 분리해주신 덕택에 테스트 메서드에서 어떤 것에 집중해야하는지 한 눈에 띄이네요 😄 👍

Hyeon9mak commented on src/main/java/roomescape/service/ReservationService.java:
```diff
@@ -16,27 +21,59 @@ public class ReservationService {
+			throw new RoomEscapeException(ErrorCode.PAST_RESERVATION);
+		}
+
+		if (this.reservationRepository.isDuplicateReservation(request)) {
+			throw new RoomEscapeException(ErrorCode.DUPLICATE_RESERVATION);
+		}
```
중복 예약이 존재하는지도 꼼꼼하게 확인해주셨군요 👍

Hyeon9mak commented on src/main/java/roomescape/repository/ReservationRepository.java:
```diff
@@ -39,32 +42,47 @@ public List<Reservation> findAll() {
 		return reservation;
 	}
 
+	public boolean isDuplicateReservation(ReservationRequest request) {
+		String sql = "SELECT COUNT(*) FROM reservation WHERE date = ? AND time_id = ?";
+		int count = this.jdbcTemplate.queryForObject(sql, Integer.class, request.date(), request.timeId());
+		return count > 0;
+	}
+
```
layered architecture 를 이용하는 이유는 유지보수 경계를 명확히 나누고, 
요구사항이 전달될 때 다른 layer 은 신경쓰지 않기 위함이라고 생각합니다!
현재 `ReservationRequest` 가 controller - service - repository 까지 다이렉트로 전달되는 형태로 보이네요.
(persistence layer 에서는 내가 수행할 동작이 무엇인지가 중요할 뿐, 어떤 요청에 의한 것인지 신경쓸 필요가 없기도 하죠!)
유근님은 어떻게 생각하시나요?

eottabom replied:
저도 리뷰어님과 동일하게 생각해요~
비즈니스 로직으로 예외를 처리하기 위해서는 모든 request 를 전달하기 보다는 필요한 부분만 전달하는게 좋은 것 같아요

https://github.com/next-step/spring-roomescape-reservation/pull/70/commits/853f04139e66ed05fddce6f9d66e84726b2b568f#diff-a08e170bf90a7f2fae99eafa05dfd10de6889b286503b78852bda8c78832cb3dR74

`ReservationRequest` 를 다이렉트로 전달하기 보다는 검증해야 하는 부분만 매개변수로 추가했어요

이 방법 말고 다른 방법도 있으면 가이드 부탁드려요~

Hyeon9mak commented on src/main/java/roomescape/controller/dto/ReservationTimeRequest.java:
```diff
@@ -1,4 +1,5 @@
 package roomescape.controller.dto;
 
 public record ReservationTimeRequest(String startAt) {
+
 }
```
<img width="551" alt="image" src="https://github.com/next-step/spring-roomescape-reservation/assets/37354145/4f19dd46-f485-4a1a-844e-ee20d09674f6">

시간, 테마에 대하여 아무런 정보가 입력되지 않아도 생성이 가능한 것 같아요 😱 
이에 관한 검증이 추가되어야 겠네요!

eottabom replied:
해당 부분 수정했어요~

Hyeon9mak commented on src/main/resources/templates/admin/reservartion-new.html:
```diff
@@ -0,0 +1,61 @@
+<!DOCTYPE html>
+<html lang="ko">
+<head>
+    <meta charset="UTF-8">
+    <meta name="viewport" content="width=device-width, initial-scale=1.0">
+    <title>방탈출 어드민</title>
+    <!-- Bootstrap CSS -->
+    <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.5.2/css/bootstrap.min.css">
```
시간 관리 페이지(time.html)의 navigation bar 도 관리가 필요할 것 같아요!

eottabom replied:
https://github.com/next-step/spring-roomescape-reservation/pull/70/commits/853f04139e66ed05fddce6f9d66e84726b2b568f#diff-2afd5a74bf935642a25620f8d9ee30e4e62c1eb5e965924db4d97529e6f07685

추가했어요~

Hyeon9mak commented on src/main/java/roomescape/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,56 @@
+
+	private static final Logger logger = LoggerFactory.getLogger(GlobalExceptionHandler.class);
+
+	@ExceptionHandler(RoomEscapeException.class)
+	public ProblemDetail handleRoomEscapeExceptions(RoomEscapeException ex, WebRequest request) {
+		ErrorCode errorCode = ex.getErrorCode();
+		HttpStatus status = getStatusForErrorCode(errorCode);
+
+		ProblemDetail problemDetail = ProblemDetail.forStatusAndDetail(status, errorCode.getMessage());
+		problemDetail.setTitle("API Error");
+		problemDetail.setInstance(URI.create(request.getDescription(false)));
+		problemDetail.setProperty("errorCode", errorCode.getCode());
+
+		logger.error("Error occurred: {}", problemDetail, ex);
+
```
<img width="1782" alt="image" src="https://github.com/next-step/spring-roomescape-reservation/assets/37354145/5d7ae3e9-7fbf-460a-b8e5-23bf188796c6">

사용자가 어떤 연유로 예약 생성에 실패했는지 인지하고, 다음 행동을 결정할 수 있도록 메세지를 전달해주는 건 어떨까요? 😉

Hyeon9mak replied:
개인적인 궁금함으로 질문이 있습니다!
REST API 규약을 지켜주신 덕분에 status code 가 이미 별도로 잘 전달될텐데,
별도로 response 에 status code 를 명세해주신 이유가 있을까요?

eottabom replied:
유효하지 않는 값으로 입력했을 때 예외 처리 추가했어요~
https://github.com/next-step/spring-roomescape-reservation/pull/70/commits/853f04139e66ed05fddce6f9d66e84726b2b568f#diff-ce52ea65d195a4a344d7e0955ee2036bf04e2c78e27437a42ff851d77f03bd97R27

--- 
말씀 주신 것처럼 REST API 규약 때문에 status code 가 별도로 전달될거예요
별도로 response 에 Status Code 를 명세한 이유는
개발자 간의 의사소통을 위한 목적...? 이 가장 클 거 같은데, 굳이 사용하지 않아도 되긴 할 것 같아요
약간의 제 습관일 수도 있는데, 이 부분에 대해서 다른 의견이 있으시면 편하게 말씀주세요~

Hyeon9mak reviewed (changes requested):
안녕하세요 유근님~! 이번엔 실제 페이지 동작, 예외와 관련하여 코멘트 몇 가지 남겼습니다!

eottabom committed:
STEP2: 예외처리 및 리팩터링

- fix: time.html 에 nav bar 추가
- fix: handleHttpMessageNotReadable 추가
- fix: isDuplicateReservation 수정
- fix: 예외 처리 추가

eottabom committed:
STEP2: 예외처리 및 리팩터링

- fix: 에러코드 개선
- fix: 테마 예외처리 추가

eottabom committed:
STEP2: 예외처리 및 리팩터링

- fix: 에러 응답 코드 추가

eottabom committed:
STEP2: 예외처리 및 리팩터링

- fix: 예외처리 누락 파일 추가

eottabom committed:
STEP2: 예외처리 및 리팩터링

- test: 불필요한 import 제거

Hyeon9mak commented on src/main/java/roomescape/repository/ReservationRepository.java:
```diff
@@ -39,32 +42,47 @@ public List<Reservation> findAll() {
+		String sql = "SELECT COUNT(*) FROM reservation WHERE date = ? AND time_id = ?";
+		int count = this.jdbcTemplate.queryForObject(sql, Integer.class, date, timeId);
+		return count > 0;
+	}
+
 	public void delete(long id) {
 		String sql = "DELETE FROM reservation WHERE id = ?";
-		int result = this.jdbcTemplate.update(sql, id);
-
-		if (result == 0) {
-			throw new RuntimeException("No data found for ID: " + id);
-		}
+		this.jdbcTemplate.update(sql, id);
```
😉 👍

Hyeon9mak commented on src/main/java/roomescape/controller/dto/ReservationTimeRequest.java:
```diff
@@ -1,4 +1,15 @@
+
+import org.springframework.util.ObjectUtils;
+
 public record ReservationTimeRequest(String startAt) {
+
+	public static void validateReservationTime(ReservationTimeRequest request) {
+		if (ObjectUtils.isEmpty(request.startAt())) {
+			throw new RoomEscapeException(ErrorCode.INVALID_TIME);
+		}
```
점차 presentation 관련 검증이 많아지므로, spring validaiton 라이브러리를 활용해도 좋겠네요 😄 
https://www.baeldung.com/spring-boot-bean-validation

Hyeon9mak reviewed (approved):
안녕하세요 유근님! 피드백 빠르게 잘 반영해주셨습니다!
이미 완성도가 높았기 때문에 더 이상 코멘트 드릴 부분이 없는 것 같네요 😄 
그동안 드린 코멘트들이 도움이 되셨을까요?!
첫 번째 미션 고생 많으셨습니다!

eottabom commented:
@Hyeon9mak
깊게 생각하지 않고 관습적(?) 으로 넘어가던 부분에 대해서 여러 방면으로 생각 할 수 있어서 좋았고, 저도 더 찾아보면서 학습하게 되서, 도움이 많이 되었어요~
저도 시각이 넓어지는거 같아요!
